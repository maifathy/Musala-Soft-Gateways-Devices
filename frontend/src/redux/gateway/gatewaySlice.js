import { createSlice } from '@reduxjs/toolkit';

const initialState = {
  gateways: []
};
const gatewaysStr = 'gateways';
export const gatewaySlice = createSlice({
  name: 'gateway',
  initialState,
  reducers: {
    setGateways: (state, action) => {
      const updatedState = { ...state };
      updatedState[gatewaysStr] = action.payload;
      return updatedState;
    },
    removeGateway: (state, action) => {
      state.gateways.filter((gateway) => gateway.id !== action.payload);
    },
    appendGateway: (state, action) => {
      state.gateways.push(action.payload);
    },
    removeDevice: (state, action) => {
      state.gateways.filter((gateway) => {
        return gateway.devices.map((device, index) => {
          if (device._id === action.payload) {
            gateway.devices.splice(index, 1);
          }
          return gateway.devices;
        });
      });
    },
    appendDevice: (state, action) => {
      state.gateways.filter((gateway) => {
        if (gateway._id === action.payload.gatewayId) {
          gateway.devices.push(action.payload.device);
        }
        return gateway.devices;
      });
    }
  }
});

// Action creators are generated for each case reducer function
export const {
  setGateways, removeGateway, appendGateway, removeDevice, appendDevice
} = gatewaySlice.actions;

export default gatewaySlice.reducer;
