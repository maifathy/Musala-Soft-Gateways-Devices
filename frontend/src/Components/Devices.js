import React, { useState, useCallback, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { useDispatch, useSelector } from 'react-redux';
import { appendDevice, removeDevice } from '../redux/gateway/gatewaySlice.js';
import { addDevice, deleteDevice } from '../utils/api.js';
import styled from 'styled-components';

const Devices = (props) => {
  const [vendor, setVendor] = useState('');
  const [status, setStatus] = useState('');
  const [message, setMessage] = useState('');
  const dispatch = useDispatch();
  const devices = useSelector((state) => {
    var gateway = state.gateway.gateways.filter((item) => {
      return item._id === props.gatewayId;
    });
    return gateway[0] ? gateway[0].devices : [];
  });
  const isUnmountedRef = useRef(false);

  useEffect(() => {
    return () => {
      // cleanup function will occur when component unmounted.
      isUnmountedRef.current = true;
    } 
  },[]); 

  useEffect(() => {
    if(message === '')
      return;
    let timerId = setTimeout(() => setMessage(''), 7000);
    // cleanup the timer when component unmout
    return () => clearTimeout(timerId);
  }, [message]);

  const delDevice = (e, id) => {
    //setMessage('');
    e.preventDefault();
    deleteDevice(id)
      .then((data) => {
        if (data.status === 200) {
          dispatch(removeDevice(id));
        }
        if(!isUnmountedRef.current) setMessage(data.message);
      })
      .catch((err) => { if(!isUnmountedRef.current) setMessage(err.message); });
  };

  const newDevice = useCallback((e) => {
    //setMessage('');
    e.preventDefault();
    addDevice(vendor, status, props.gatewayId)
      .then((device) => {
        if (device.status === 200) {
          dispatch(appendDevice({ device: device.device, gatewayId: props.gatewayId }));
          if(!isUnmountedRef.current) {
            setMessage(device.message);
            setVendor('');
            setStatus('');
          }
        }
      })
      .catch((err) => { if(!isUnmountedRef.current) setMessage(err.message); });
  }, [vendor, status, props.gatewayId]);

  return (
    <>
      {message && <p data-testid="message">{message}</p>}
      <ul>
        {devices.map(
          (device) => <ListItem key={device._id}>
            <div>
              <label>{device.vendor}</label> |
              <label>{device.status}</label> |
              <Button data-testid={`button_del_${device._id}`} type='button' style={{ width: '100px', marginLeft: '10px' }} onClick={(e) => delDevice(e, device._id)}>Remove</Button>
            </div>
          </ListItem>
        )
      }
      </ul>
      {devices.length < 10
        && <form onSubmit={newDevice} className='form' data-testid="form">
          <Input data-testid="vendor_input" type='text' value={vendor} placeholder='Vendor' onChange={(e) => setVendor(e.target.value)}/>
          <Select data-testid="status_select" value={status} onChange={(e) => setStatus(e.target.value)}>
            <option value=''>Status</option>
            <option value='Online'>Online</option>
            <option value='Offline'>Offline</option>
          </Select>
          <div>
            <Button data-testid="submit" type='submit' disabled={vendor === '' || status === ''}>Submit</Button>
          </div>
        </form>
      }
      </>
  );
};
Devices.propTypes = {
  gatewayId: PropTypes.number
};

export default Devices;

const Input = styled.input`
  border: 1px solid #a6a6ca;
  border-radius: 2px;
  font-size: 14px;
  color: navy;
  height: 30px;
  :focus {
    outline: none;
    background-color: rgb(215 215 229);
  }
`;

const Select = styled.select`
  border: 1px solid #a6a6ca;
  border-radius: 2px;
  font-size: 14px;
  height: 34px;
  :focus {
    outline: none;
    background-color: rgb(215 215 229);
  }
`;

const Button = styled.button`
  margin-top: 10px;
  padding: 5px;
  width: 200px;
  border: 1px solid "#7620ff";
  background-color: "#7620ff";
  outline: none;
  color: "#fff";
  :hover {
    background-color: "#352878";
  }
`;

const ListItem = styled.li`
  background-color: aliceblue;
  margin: 3px;
  padding: 5px;
  list-style-type: square;
`;