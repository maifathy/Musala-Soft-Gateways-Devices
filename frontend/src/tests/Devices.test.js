/* eslint-disable no-undef */
import React from 'react';
import Enzyme, { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { render as renderLibrary, fireEvent, screen, cleanup } from '@testing-library/react';
import { waitFor } from "@testing-library/dom";
import { store } from '../app/store.js';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import history from '../utils/history.js';
import { addDevice, addGateway, getGateway, clearCollections } from '../utils/api.js';
import Devices from '../Components/Devices.js';
import { appendGateway } from '../redux/gateway/gatewaySlice.js';

Enzyme.configure({ adapter: new Adapter() })

it('Devices Snapshot - renders correctly enzyme', () => {
    const wrapper = mount(<Provider store={store}>  <BrowserRouter history={history} location={history.location} navigator={history}> <Devices gatewayId={0}/> </BrowserRouter> </Provider>);

    expect(toJson(wrapper)).toMatchSnapshot();
});

beforeEach(() => { jest.setTimeout(80000) });
afterEach(() => { 
  clearCollections().then(() => {
    mongoose.connection.close();
    cleanup(); 
  });
});

// Delete existing Device
it('Delete existing Device!', async () => {
  // insert default Gateway
  let gatewayId = 0;
  let deviceId = "";
  await addGateway("Gateway2", "127.0.0.10")
    .then((response) => {
      if(response.status === 400)
        return;
      gatewayId = response.gateway._id;
  });
  
  await addDevice("Device2", "Offline", gatewayId)
  .then((response) => {
    if(response.status === 400)
      return;
    deviceId = response.device._id;
  });

  await getGateway(gatewayId)
  .then(async(response) => {
    if(response === 404)
      return;
    store.dispatch(appendGateway(response.gateway));
  })
  .catch((error) => {
    console.log("getGateway Error: ", error);
  });
  const { getByTestId } = renderLibrary(<Provider store={store}>  <BrowserRouter history={history} location={history.location} navigator={history}> <Devices gatewayId={gatewayId}/> </BrowserRouter> </Provider>);
  
  fireEvent.click(getByTestId('button_del_' + deviceId));
  await waitFor(() =>
    expect(screen.queryAllByTestId(`button_del_${deviceId}`).length).toEqual(1),
    //expect(getByTestId('message')).toHaveTextContent('Device is successfully removed')
  );
});

// Add New Device
it('Add New Device!', async () => {
    // insert default Gateway
    let gatewayId = 0;
    await addGateway("Default Gateway", "127.0.0.1")
    .then((response) => {
      if(response.status === 400)
        return;
      gatewayId = response.gateway._id;
    })
    const { getByText, getByTestId } = renderLibrary(<Provider store={store}>  <BrowserRouter history={history} location={history.location} navigator={history}> <Devices gatewayId={gatewayId}/> </BrowserRouter> </Provider>);

    expect(getByText('Submit')).toBeDisabled();

    fireEvent.change(getByTestId('vendor_input'), { target: { value: 'Device1'}});
    fireEvent.change(getByTestId('status_select'), { target: { value: 'Online'}});

    expect(getByText('Submit')).not.toBeDisabled();

    fireEvent.click(getByText('Submit')); // alternative Submit
    await waitFor(() =>
      //expect(getByTestId('message')).toHaveTextContent('Device is successfully added'),
      expect(getByTestId('vendor_input')).toBeEmptyDOMElement(),
      expect(getByTestId('status_select')).toHaveTextContent("Status")
    );
});