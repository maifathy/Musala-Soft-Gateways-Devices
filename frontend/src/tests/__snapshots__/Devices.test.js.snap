// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Devices Snapshot - renders correctly enzyme 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
    
  <BrowserRouter
    history={
      Object {
        "action": "POP",
        "back": [Function],
        "block": [Function],
        "createHref": [Function],
        "forward": [Function],
        "go": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    location={
      Object {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      }
    }
    navigator={
      Object {
        "action": "POP",
        "back": [Function],
        "block": [Function],
        "createHref": [Function],
        "forward": [Function],
        "go": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Router
      location={
        Object {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        }
      }
      navigationType="POP"
      navigator={
        Object {
          "action": "POP",
          "createHref": [Function],
          "createURL": [Function],
          "encodeLocation": [Function],
          "go": [Function],
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "default",
            "pathname": "/",
            "search": "",
            "state": null,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
       
      <Devices
        gatewayId={0}
      >
        <ul />
        <form
          className="form"
          data-testid="form"
          onSubmit={[Function]}
        >
          <styled.input
            data-testid="vendor_input"
            onChange={[Function]}
            placeholder="Vendor"
            type="text"
            value=""
          >
            <input
              className="sc-bgqQcB cAbwyu"
              data-testid="vendor_input"
              onChange={[Function]}
              placeholder="Vendor"
              type="text"
              value=""
            />
          </styled.input>
          <styled.select
            data-testid="status_select"
            onChange={[Function]}
            value=""
          >
            <select
              className="sc-gTRrQi gURgaj"
              data-testid="status_select"
              onChange={[Function]}
              value=""
            >
              <option
                value=""
              >
                Status
              </option>
              <option
                value="Online"
              >
                Online
              </option>
              <option
                value="Offline"
              >
                Offline
              </option>
            </select>
          </styled.select>
          <div>
            <styled.button
              data-testid="submit"
              disabled={true}
              type="submit"
            >
              <button
                className="sc-ewnqHT gbhdnH"
                data-testid="submit"
                disabled={true}
                type="submit"
              >
                Submit
              </button>
            </styled.button>
          </div>
        </form>
      </Devices>
       
    </Router>
  </BrowserRouter>
   
</Provider>
`;
